keps/sig-network/4559-kubelet-probes/README.md

```diff
@@ -0,0 +1,1037 @@
+  (e.g., [multi-tenant scenarios]), and in some of these cases, people
+  have wanted to reuse the same pod IPs for different pods. Although
+  it would be possible to deploy a Kubernetes cluster with a CNI
+  plugin, service proxy and NetworkPolicy implementation that properly
+  handle non-unique pod IPs, there is currently no way to make
+  network-based pod probes work correctly in such an environment.
+
+  The proposed [Multi-Network feature] raises similar issues, in that
+  in addition to the possibility of networks with overlapping IPs, it
+  also potentially involves networks that cannot be reached from the
+  host-network namespace.
+
+- **Dual-Stack Pods**: Probes are currently defined to use the pod's
+  primary IP family, but there are cases where you might want to do an
+  IPv6 probe in an IPv4-primary dual-stack cluster. ([kubernetes
+  #101324])
+
+- **Unclear Semantics for Network Probes**: It was never defined
+  whether probes were intended to check "is the server process in the
+  pod running?" or "is the pod reachable over the network?" or both.
+  While many people would like them to mean the latter, this has
+  turned out to not really work in many cases. (In particular, the
+  fact that kubelet probes are exempt from NetworkPolicy means that a
+  probe may succeed even when no other pods would be able to reach the
+  pod because [the NetworkPolicy has not yet been fully programmed].)
+  This led to the creation of [readiness gates] to provide better pod
+  readiness information.
+
+- **Exec Probe Performance (?)**: [Exec probes are much, much slower than
+  network probes]. Apparently this slowness is a somewhat inevitable
+  side effect of how exec probes work in an OCI runtime context?
+  Possibly because it requires making multiple synchronous calls?
+  Anyway, it's possible this could be improved.
+
+- **Network Probe Performance (?)**: In the past, there have been problems
+  with probes consuming unexpectedly large amounts of kubelet resources
+  ([kubernetes #89898], "Sometime Liveness/Readiness Probes fail because
+  of net/http: request canceled while waiting for connection
+  (Client.Timeout exceeded while awaiting headers)"). This has been
+  mitigated to some extent ([kubernetes #115143])... it's not clear if
+  this is still considered a problem.
+
+See further discussion in [kubernetes #102613].
+
+#### RESOLVED Summary discussions (preserving for history)
+- Figure out if there are any concerns with non-namespace-based Linux
+  runtimes (kata, kubevirt, etc): There are no concerns here because both kata containers and [kubevirt] use network namespaces. They expect the CNI to set up the pod networking. Then their agent creates the required networking infrastructure inside this pod network namespace to connect to the VM interface
+
+#### UNRESOLVED Summary discussions
+
+- Figure out details of exec probe performance problem.
+
+- Figure out if network probe performance is still a problem.
```

Comment:
@mikebrow you may be interested in this KEP, it may help to  implement subseconds probes https://github.com/kubernetes/enhancements/pull/3067